Mathlib/LinearAlgebra/RootSystem/Basic.lean

```diff
@@ -315,13 +207,43 @@ protected lemma ext [CharZero R] [NoZeroSMulDivisors R M]
   apply P₁.bijectiveRight.injective
   apply Dual.eq_of_preReflection_mapsTo (P₁.ne_zero i) (finite_range P₁.root) P₁.span_eq_top
   · exact hr ▸ he ▸ P₂.coroot_root_two i
-  · exact hr ▸ he ▸ P₂.mapsTo_preReflection_root i
+  · exact hr ▸ he ▸ P₂.mapsTo_reflection_root i
   · exact P₁.coroot_root_two i
-  · exact P₁.mapsTo_preReflection_root i
+  · exact P₁.mapsTo_reflection_root i
+
+private lemma coroot_eq_coreflection_of_root_eq_of_span_eq_top [CharZero R] [NoZeroSMulDivisors R M]
+    (p : PerfectPairing R M N)
+    (root : ι ↪ M)
+    (coroot : ι ↪ N)
+    (hp : ∀ i, p.toLin (root i) (coroot i) = 2)
+    (hs : ∀ i, MapsTo (preReflection (root i) (p.toLin.flip (coroot i))) (range root) (range root))
+    (hsp : span R (range root) = ⊤)
+    {i j k : ι} (hk : root k = preReflection (root i) (p.toLin.flip (coroot i)) (root j)) :
+    coroot k = preReflection (coroot i) (p.toLin (root i)) (coroot j) := by
+  set α := root i
+  set β := root j
+  set α' := coroot i
+  set β' := coroot j
+  set sα := preReflection α (p.toLin.flip α')
+  set sβ := preReflection β (p.toLin.flip β')
+  let sα' := preReflection α' (p.toLin α)
+  have hij : preReflection (sα β) (p.toLin.flip (sα' β')) = sα ∘ₗ sβ ∘ₗ sα := by
+    ext
+    have hpi : (p.toLin.flip (coroot i)) (root i) = 2 := by rw [LinearMap.flip_apply, hp i]
+    simp [α, β, α', β', sα, sβ, sα', ← preReflection_preReflection β (p.toLin.flip β') hpi,
+      preReflection_apply] -- v4.7.0-rc1 issues
+  have hk₀ : root k ≠ 0 := fun h ↦ by simpa [h] using hp k
+  apply p.bijectiveRight.injective
+  apply Dual.eq_of_preReflection_mapsTo hk₀ (finite_range root) hsp (hp k) (hs k)
+  · simp [α, β, α', β', sα, sβ, sα', hk, preReflection_apply, hp i, hp j, mul_two,
+      mul_comm (p.toLin α β')]
+    ring -- v4.7.0-rc1 issues
+  · rw [hk, hij]
+    exact (hs i).comp <| (hs j).comp (hs i)
 
-/-- In characteristic zero if there is no torsion, to check that a family of roots form a root
-system, we do not need to check that the coroots are stable under reflections since this follows
-from the corresponding property for the roots. -/
+/-- In characteristic zero if there is no torsion, to check that a finite family of roots form a
+root system, we do not need to check that the coroots are stable under reflections since this
+follows from the corresponding property for the roots. -/
 def mk' [CharZero R] [NoZeroSMulDivisors R M]
```

Comment:
Would it be possible to define this using
```
  [.......] where
  span_eq_top := hsp
  toRootPairing := RootPairing.mk' p root coroot hp hs <| by
    intro i
    sorry
```
or will that sorry be really hard to fill in?

Reply:
Done.